src/ifcgeom/taxonomy.h

```diff
@@ -125,6 +125,41 @@ struct item {
 	uint32_t identity() const { return identity_; }
 };
 
+struct implicit_item : public item {
+	DECLARE_PTR(implicit_item)
+	
+	virtual item::ptr evaluate() const = 0;
+};
+
+struct piecewise_function : public implicit_item {
+	DECLARE_PTR(piecewise_function)
+
+	std::vector<std::pair<double, std::function<Eigen::VectorXd(double u)>>> spans;
+
+	void print(std::ostream& o, int indent = 0) const {
+		o << "piecewise_function" << std::endl;
+	}
+
+	virtual piecewise_function* clone_() const { return new piecewise_function(*this); }
+	virtual kinds kind() const { return PIECEWISE_FUNCTION; }
+
+	virtual size_t calc_hash() const {
+		auto v = std::make_tuple(static_cast<size_t>(PIECEWISE_FUNCTION), 0);
+		return boost::hash<decltype(v)>{}(v);
+	}
+
+	virtual item::ptr evaluate() const;
+
+	Eigen::VectorXd evaluate(double u) const {
+		// @todo optimize, assume monotonic evaluation and store last evaluated segment?
```

Comment:
If span had start and end instead of length, this method could be optimized something like:

```
static std::tuple<double, double, std::function<Eigen::VectorXd(double u)>>* last_span = nullptr;
if(last_span)
{
  // there was a previous evaluated span
   auto& [start,end,fn] = *last_span;
   if(start <= u && u < end)
      return fn(u - start); // u is in the range for the previous span, evaluate it
}

   // no last span, or u is not in the range of the last span.... find the span that covers u
   auto iter = std::find_if(spans.begin(),spans.end(),[u](auto& s){return u <= s.start && u < s.end;});
   if(iter == spans.end()) { // deal with special case of u = s.end for last segment or error if u is not in the range of any spans}
   else {
      last_span = &(*iter);
      auto& [start,end,fn] = *iter;
      return fn(u - start);
   }
```

Reply:
Yes, I can add this to my pr. Good point about `static`. I'll address it appropriately.

Reply:
@aothms @civilx64 I'm have a little trouble getting this working. Since the concept is to be more efficient, I think it is more important to focus on getting the geometry we need to advance IFCOS. I can return to making this more efficient at a later date.

Reply:
Absolutely agree. Same for things like hashing and all. I can work on that independently later on.